### ISSUE TITLE
[JENKINS-65308] Jenkins.trimLabels gets increasingly slower as number of nodes and labels increase

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#8979]
<p>We've been trying to track down some issues we've been seeing around Queue lock<br/>
contention on one of our Jenkins clusters. The lock contention manifests in<br/>
both UI instability/slowness and failures with REST API calls to add, update or<br/>
remove nodes. We use the Swarm plugin on the primary and swarm-client (version<br/>
3.24) on the agents to connect to the primary. The REST API failures aren't due<br/>
to exceptions from Jenkins, but to the API calls exceeding the configured<br/>
proxy_read_timeout (180s) for the nginx instance we have in front of Jenkins.<br/>
That manifests in the swarm-client process on the agents receiving a 504 from<br/>
nginx since Jenkins didn't respond in time.</p>

<p>Thread dumps gathered during periods of instability show that hundreds of<br/>
threads are waiting for the Queue lock to be able to add, update or remove<br/>
a node.</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>"Handling POST /plugin/swarm/createSlave from 10.224.1.234 : Jetty (winstone)-1218487" #1218487 prio=5 os_prio=0 tid=0x00007f4732c7f800 nid=0x2f96 waiting on condition [0x00007f3c6b5f2000]
   java.lang.Thread.State: WAITING (parking)
       at sun.misc.Unsafe.park(Native Method)
       - parking to wait for  <0x00007f3f117ca288> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
       at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
       at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
       at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:870)
       at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
       at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:209)
       at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:285)
       at hudson.model.Queue._withLock(Queue.java:1441)
[snip..]
</pre>
</div></div>

<p>The vast majority of the time, the thread holding the Queue lock during each<br/>
thread dump is performing operations within the Jenkins.trimLabels method as<br/>
part of adding, updating or removing a node.</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>"Handling POST /plugin/swarm/createSlave from 10.240.81.215 : Jetty (winstone)-1218362" #1218362 prio=5 os_prio=0 tid=0x00007f47265fd000 nid=0x2e11 runnable [0x00007f3c85ceb000]
   java.lang.Thread.State: RUNNABLE
at hudson.util.QuotedStringTokenizer.hasMoreTokens(QuotedStringTokenizer.java:184)
at hudson.model.Label.parse(Label.java:585)
at hudson.model.Node.getAssignedLabels(Node.java:303)
at hudson.model.Label.matches(Label.java:196)
at hudson.model.Label.getNodes(Label.java:233)
at hudson.model.Label.isEmpty(Label.java:430)
at jenkins.model.Jenkins.trimLabels(Jenkins.java:2201)
at jenkins.model.Nodes$4.call(Nodes.java:214)
at jenkins.model.Nodes$4.call(Nodes.java:210)
at hudson.model.Queue._withLock(Queue.java:1443)
at hudson.model.Queue.withLock(Queue.java:1304)
at jenkins.model.Nodes.updateNode(Nodes.java:210)
at jenkins.model.Jenkins.updateNode(Jenkins.java:2176)
at hudson.model.Node.save(Node.java:139)
at hudson.model.Node.setTemporaryOfflineCause(Node.java:274)
at hudson.model.Computer.setNode(Computer.java:820)
at hudson.slaves.SlaveComputer.setNode(SlaveComputer.java:895)
at hudson.model.AbstractCIBase.updateComputer(AbstractCIBase.java:137)
at hudson.model.AbstractCIBase.access$000(AbstractCIBase.java:43)
at hudson.model.AbstractCIBase$2.run(AbstractCIBase.java:223)
at hudson.model.Queue._withLock(Queue.java:1384)
at hudson.model.Queue.withLock(Queue.java:1261)
at hudson.model.AbstractCIBase.updateComputerList(AbstractCIBase.java:206)
at jenkins.model.Jenkins.updateComputerList(Jenkins.java:1632)
at jenkins.model.Nodes$2.run(Nodes.java:151)
at hudson.model.Queue._withLock(Queue.java:1384)
at hudson.model.Queue.withLock(Queue.java:1261)
at jenkins.model.Nodes.addNode(Nodes.java:147)
at jenkins.model.Jenkins.addNode(Jenkins.java:2155)
at hudson.plugins.swarm.PluginImpl.doCreateSlave(PluginImpl.java:224)
</pre>
</div></div>

<p>I've attached a couple of archives created using the <tt>collectPerformanceData</tt><br/>
script that contain the relevant thread dumps.</p>

<p>During the aforementioned periods of instability there are between 1500-1600<br/>
unique labels and 400-500 workers, as gathered from the script console using<br/>
<tt>Jenkins.instance.labels.size()</tt> and <tt>Jenkins.instance.nodes.size()</tt>. </p>

<p>I'm able to replicate the increasing slowness using Groovy scripts that mirror<br/>
what our worker creation steps look like. I've attached both scripts.<br/>
<tt>create-workers.groovy</tt> creates the workers, <tt>remove-workers.groovy</tt> removes<br/>
them. To make it match our swarm-client workflow we create <tt>SwarmSlave</tt> agents<br/>
in the script but that detail probably doesn't matter for reproduction<br/>
purposes.</p>

<p>Creating and then removing workers with a single label is fast, as you'd<br/>
expect. Here's some snipped output for creation (full output attached as <tt>create-workers-single-label.log</tt>):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>...
uniqueLabels: 395 nodes: 393 swarm-test-392: 63ms
uniqueLabels: 396 nodes: 394 swarm-test-393: 65ms
uniqueLabels: 397 nodes: 395 swarm-test-394: 87ms
uniqueLabels: 398 nodes: 396 swarm-test-395: 62ms
uniqueLabels: 399 nodes: 397 swarm-test-396: 62ms
uniqueLabels: 400 nodes: 398 swarm-test-397: 62ms
uniqueLabels: 401 nodes: 399 swarm-test-398: 63ms
uniqueLabels: 402 nodes: 400 swarm-test-399: 63ms
uniqueLabels: 403 nodes: 401 swarm-test-400: 64ms
Total time to create 400 workers: 9183ms
</pre>
</div></div>

<p>And then the same for removal (full output attached as <tt>remove-workers-single-label.log</tt>):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>...
uniqueLabels: 10 nodes: 8 swarm-test-91: 0ms
uniqueLabels: 9 nodes: 7 swarm-test-92: 0ms
uniqueLabels: 8 nodes: 6 swarm-test-93: 1ms
uniqueLabels: 7 nodes: 5 swarm-test-94: 0ms
uniqueLabels: 6 nodes: 4 swarm-test-95: 1ms
uniqueLabels: 5 nodes: 3 swarm-test-96: 0ms
uniqueLabels: 4 nodes: 2 swarm-test-97: 1ms
uniqueLabels: 3 nodes: 1 swarm-test-98: 0ms
uniqueLabels: 1 nodes: 0 swarm-test-99: 1ms
Total time to remove 401 workers: 8675ms
</pre>
</div></div>

<p>But once you start adding more labels, thing start slowing down drastically.<br/>
Here's some snipped output for creation (full output attached as <tt>create-workers-multiple-labels.log</tt>):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>...
uniqueLabels: 809 nodes: 393 swarm-test-392: 1875ms
uniqueLabels: 811 nodes: 394 swarm-test-393: 1875ms
uniqueLabels: 813 nodes: 395 swarm-test-394: 1883ms
uniqueLabels: 815 nodes: 396 swarm-test-395: 1888ms
uniqueLabels: 817 nodes: 397 swarm-test-396: 1901ms
uniqueLabels: 819 nodes: 398 swarm-test-397: 1913ms
uniqueLabels: 821 nodes: 399 swarm-test-398: 1915ms
uniqueLabels: 823 nodes: 400 swarm-test-399: 1927ms
uniqueLabels: 825 nodes: 401 swarm-test-400: 1939ms
Total time to create 400 workers: 261866ms
</pre>
</div></div>
<p>And then the same for removal (full output attached as <tt>remove-workers-multiple-labels.log</tt>):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>...
uniqueLabels: 39 nodes: 8 swarm-test-91: 3ms
uniqueLabels: 37 nodes: 7 swarm-test-92: 2ms
uniqueLabels: 35 nodes: 6 swarm-test-93: 2ms
uniqueLabels: 33 nodes: 5 swarm-test-94: 1ms
uniqueLabels: 31 nodes: 4 swarm-test-95: 1ms
uniqueLabels: 29 nodes: 3 swarm-test-96: 0ms
uniqueLabels: 27 nodes: 2 swarm-test-97: 0ms
uniqueLabels: 25 nodes: 1 swarm-test-98: 1ms
uniqueLabels: 1 nodes: 0 swarm-test-99: 0ms
Total time to remove 401 workers: 258555ms
</pre>
</div></div>

<p>Increasing (roughly doubling it in this case) the number of unique labels makes<br/>
the same process that originally took about 9s for each operation take about<br/>
4 minutes and 20 seconds for each operation.</p>

<p>Is there some way to make <tt>Jenkins.trimLabels</tt> less expensive even in the<br/>
face of thousands of labels and hundreds of workers? To my eye it looks like<br/>
the current code path has several nested loops (outer loop over every label,<br/>
inner loop over every worker, inner loop over every parsed token from the label<br/>
tokenizer, inner loop over every char in the raw label str) which are what<br/>
contribute to the increase in execution time as the inputs get larger.</p>

**Attachments:**

- [create-workers-multiple-labels.log](https://issues.jenkins.io/secure/attachment/54419/create-workers-multiple-labels.log)
- [create-workers-single-label.log](https://issues.jenkins.io/secure/attachment/54418/create-workers-single-label.log)
- [create-workers.groovy](https://issues.jenkins.io/secure/attachment/54422/create-workers.groovy)
- [performanceData.13272-2.output.tar.gz](https://issues.jenkins.io/secure/attachment/54417/performanceData.13272-2.output.tar.gz)
- [performanceData.13272-3.output.tar.gz](https://issues.jenkins.io/secure/attachment/54416/performanceData.13272-3.output.tar.gz)
- [remove-workers-multiple-labels.log](https://issues.jenkins.io/secure/attachment/54421/remove-workers-multiple-labels.log)
- [remove-workers-single-label.log](https://issues.jenkins.io/secure/attachment/54420/remove-workers-single-label.log)
- [remove-workers.groovy](https://issues.jenkins.io/secure/attachment/54423/remove-workers.groovy)

---
<details><summary><i>Originally reported by <a title="Jonah Bull" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jonahbull">jonahbull</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-65308" target="_blank">Jenkins.trimLabels gets increasingly slower as number of nodes and labels increase</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Raihaan Shouhell" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=raihaan">raihaan</a>
<li><b>reported by</b>: jonahbull
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2021-04-19T09:23:26+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 9
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jonahbull">jonahbull</a>:</i>
<p>Really appreciate the quick response to this issue! After the linked PR was<br/>
merged I built <tt>jenkins.war</tt> from master locally and built a jdk11 Docker<br/>
image based on the that WAR file using the Makefile from the Jenkins docker<br/>
repo. Unfortunately in my tests the behavior I described looks worse and not<br/>
better after these changes.</p>

<p>Here's some snipped output for creation using latest master:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>jenkins@3fd5d3f4801e:~$ tail create-workers-with-labels.log
uniqueLabels: 811 nodes: 393 swarm-test-392: 1019ms
uniqueLabels: 813 nodes: 394 swarm-test-393: 949ms
uniqueLabels: 815 nodes: 395 swarm-test-394: 2334ms
uniqueLabels: 817 nodes: 396 swarm-test-395: 2048ms
uniqueLabels: 819 nodes: 397 swarm-test-396: 996ms
uniqueLabels: 821 nodes: 398 swarm-test-397: 972ms
uniqueLabels: 823 nodes: 399 swarm-test-398: 1036ms
uniqueLabels: 825 nodes: 400 swarm-test-399: 1096ms
uniqueLabels: 827 nodes: 401 swarm-test-400: 997ms
Total time to create 400 workers: 361632ms
</pre>
</div></div>
<p>And then the same for removal:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>jenkins@3fd5d3f4801e:~$ tail remove-workers-with-labels.log
uniqueLabels: 41 nodes: 8 swarm-test-91: 3ms
uniqueLabels: 39 nodes: 7 swarm-test-92: 2ms
uniqueLabels: 37 nodes: 6 swarm-test-93: 2ms
uniqueLabels: 35 nodes: 5 swarm-test-94: 3ms
uniqueLabels: 33 nodes: 4 swarm-test-95: 2ms
uniqueLabels: 31 nodes: 3 swarm-test-96: 2ms
uniqueLabels: 29 nodes: 2 swarm-test-97: 2ms
uniqueLabels: 27 nodes: 1 swarm-test-98: 2ms
uniqueLabels: 3 nodes: 0 swarm-test-99: 1ms
Total time to remove 401 workers: 391983ms
</pre>
</div></div>

<p>In contrast, here's the snipped output for creation using the 2.276 docker<br/>
image:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>jenkins@58c6fe073fc5:~$ tail create-workers-with-labels.log
uniqueLabels: 811 nodes: 393 swarm-test-392: 3430ms
uniqueLabels: 813 nodes: 394 swarm-test-393: 3445ms
uniqueLabels: 815 nodes: 395 swarm-test-394: 1824ms
uniqueLabels: 817 nodes: 396 swarm-test-395: 1619ms
uniqueLabels: 819 nodes: 397 swarm-test-396: 1682ms
uniqueLabels: 821 nodes: 398 swarm-test-397: 1665ms
uniqueLabels: 823 nodes: 399 swarm-test-398: 1652ms
uniqueLabels: 825 nodes: 400 swarm-test-399: 1676ms
uniqueLabels: 827 nodes: 401 swarm-test-400: 1668ms
Total time to create 400 workers: 234127ms
</pre>
</div></div>
<p>And then the same for removal:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>jenkins@58c6fe073fc5:~$ tail remove-workers-with-labels.log
uniqueLabels: 41 nodes: 8 swarm-test-91: 5ms
uniqueLabels: 39 nodes: 7 swarm-test-92: 4ms
uniqueLabels: 37 nodes: 6 swarm-test-93: 4ms
uniqueLabels: 35 nodes: 5 swarm-test-94: 3ms
uniqueLabels: 33 nodes: 4 swarm-test-95: 2ms
uniqueLabels: 31 nodes: 3 swarm-test-96: 2ms
uniqueLabels: 29 nodes: 2 swarm-test-97: 2ms
uniqueLabels: 27 nodes: 1 swarm-test-98: 2ms
uniqueLabels: 3 nodes: 0 swarm-test-99: 1ms
Total time to remove 401 workers: 262418ms
</pre>
</div></div>

<p>This is all using the groovy scripts attached to this issue. I'll dig some more<br/>
tomorrow and see if I can provide some further data.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=raihaan">raihaan</a>:</i>
<p>Hey Jonah thanks for the feedback, I'll probably write another PR which you can hopefully test.</p>

<p>In my limited testing, this improved things but it doesn't seem to have helped you.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=raihaan">raihaan</a>:</i>
<p>Hey <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jonahbull" class="user-hover" rel="jonahbull">jonahbull</a> if you get the chance could you try out the newer PR.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jonahbull">jonahbull</a>:</i>
<p>Hey Raihaan, thanks again for the quick work on these PRs! I tested the newer PR this morning and got not-very-improved results, which didn't make any sense to me. I slowly realized that my timing/reporting code in the worker creation loop was probably what was dominating the time now since it has to get the number of labels and nodes to print each iteration. I commented the timing code in the worker creation loop, leaving just the measurement of the total time to create/remove all the workers. That seems to have done the trick as far as reporting accurate results. So the first PR was actually a massive improvement from our baseline on 2.276 and the second PR is even better still!</p>

<p>2.276:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Total time to create 400 workers: 240958ms
Total time to remove 401 workers: 255528ms
</pre>
</div></div>

<p>PR-5402 (first PR):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Total time to create 400 workers: 2903ms
Total time to remove 401 workers: 1871ms
</pre>
</div></div>

<p>PR-5412 (second PR):</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Total time to create 400 workers: 1637ms
Total time to remove 401 workers: 1441ms
</pre>
</div></div>

<p>Personally I find the implementation of the second PR clearer so I'd love to see that one land as well. I'm not sure if there are any concerns over whether the second PR changes the contract of the <tt>trimLabels</tt> method significantly, hopefully folks who know that area of the code better will chime in on that aspect.</p>
